test/puppetlabs/trapperkeeper/testutils/logging.clj

```diff
@@ -1,28 +1,321 @@
 (ns puppetlabs.trapperkeeper.testutils.logging
-  (:import (java.util.regex Pattern)
-           (org.slf4j LoggerFactory Logger)
-           (ch.qos.logback.core AppenderBase))
-  (:require [clojure.tools.logging.impl :as impl]
-            [clojure.tools.logging :refer [*logger-factory*]]
-            [clojure.test :refer [assert-expr]]
-            [puppetlabs.trapperkeeper.logging :as pl-log]))
+  (:import
+   [ch.qos.logback.classic Level Logger]
+   [ch.qos.logback.classic.encoder PatternLayoutEncoder]
+   [ch.qos.logback.core Appender FileAppender]
+   [ch.qos.logback.core.spi LifeCycle]
+   (ch.qos.logback.core AppenderBase)
+   [org.slf4j LoggerFactory]
+   [java.util.regex Pattern])
+  (:require
+   [clojure.set :as set]
+   [clojure.test]
+   [clojure.tools.logging.impl :as impl]
+   [clojure.tools.logging :refer [*logger-factory*]]
+   [me.raynes.fs :as fs]
+   [puppetlabs.kitchensink.core :as kitchensink]
+   [puppetlabs.trapperkeeper.logging :as pl-log
+    :refer [root-logger-name]]
+   [schema.core :as s]))
+
+;; Note that the logging configuration is a global resource, so
+;; simultaneous calls to reset-logging, configure-logging!, etc. may
+;; interfere with many of these calls.
+
+(def ^:private keyword-levels
+  {:trace Level/TRACE
+   :debug Level/DEBUG
+   :info Level/INFO
+   :warn Level/WARN
+   :error Level/ERROR})
+
+(def ^:private level-keywords (set/map-invert keyword-levels))
+
+(def ^:private levels (set (keys keyword-levels)))
+
+(defn event->map [event]
+  "Returns {:logger name :level lvl :exception throwable :message msg}
+  for the given event.  Note that this does not convert any nil
+  messages to \"\"."
+  (merge
+   {:logger (.getLoggerName event)
+    :level  (level-keywords (.getLevel event))
+    :message (.getMessage event)
+    :exception (.getThrowableProxy event)}))
+
+;; Perhaps make with-started public in kitchensink or some other namespace?
+
+(defn- expand-with-started
+  [[name init & bindings] body]
+  (if (seq bindings)
+    `(let [~name ~init]
+       (try
+         ~(expand-with-started bindings body)
+         (finally (.stop ~name))))
+    `(let [~name ~init]
+       (try
+         ~@body
+         (finally (.stop ~name))))))
+
+(defmacro ^:private with-started
+  "Ensures that if a given name's init form executes without throwing
+  an exception, (.stop name) will be called before returning from
+  with-started.  It is the responsibility of the init form to make
+  sure the object has been started.  This macro behaves like
+  with-open, but with respect to .stop instead of .close."
+  [bindings & body]
+  ;; We call out to a function because Clojure currently chokes on a
+  ;; recursive private macro.
```

Comment:
FYI I am actually trying to review this, not just leave hilarious comments.  Just keep getting interrupted.